### PR TITLE
mrc-2212 - Include metadata when pushing file to ADR

### PR DESCRIPTION
**Pull request by carbowizzy**
## Description

This PR adds metadata for description field which will be pushed to the ADR against the corresponding resource (This version can not be edited by users before pushing, thus the text box is not being displayed/implemented)

https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-2367

## Type of version change
_Delete as appropriate_

Major - Minor - Patches - None

## Checklist

- [x] I have incremented version number, or version needs no increment
- [x] The build passed successfully, or failed because of ADR tests

### CODE REVIEW

File: src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/ADRController.kt [hunk 1]
```diff
@@ -167,20 +167,19 @@ class ADRController(private val encryption: Encryption,
...
     {
         // 1. Download relevant artefact from hintr
         val artefact = when (resourceType)
         {
-            appProperties.adrOutputZipSchema -> Pair(apiClient.downloadSpectrum(modelCalibrateId),
-                    "Naomi model outputs")
-            appProperties.adrOutputSummarySchema -> Pair(apiClient.downloadSummary(modelCalibrateId),
-                    "Naomi summary report")
+            appProperties.adrOutputZipSchema -> Pair(apiClient.downloadSpectrum(modelCalibrateId), description)
+            appProperties.adrOutputSummarySchema -> Pair(apiClient.downloadSummary(modelCalibrateId), description)
```
Comment (mwoodbri) [inline src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/ADRController.kt:182]: I think you can get rid of these `Pair`s - they're an unnecessary level of indirection if description is in common:
```suggestion
            appProperties.adrOutputZipSchema -> apiClient.downloadSpectrum(modelCalibrateId)
            appProperties.adrOutputSummarySchema -> apiClient.downloadSummary(modelCalibrateId)
```

File: src/app/static/src/app/store/adr/actions.ts [hunk 2]
```diff
@@ -7,6 +7,7 @@ import {ADRMutation} from "./mutations";
 import {constructUploadFile, datasetFromMetadata} from "../../utils";
 import {Organization, UploadFile, Dict} from "../../types";
 import {BaselineMutation} from "../baseline/mutations";
+import {rootState} from "../../../tests/integration/integrationTest";
```
Comment (mwoodbri) [inline src/app/static/src/app/store/adr/actions.ts:10]: This should be removed

File: src/app/static/src/app/store/adr/actions.ts [hunk 3]
```diff
@@ -143,20 +144,28 @@ export const actions: ActionTree<ADRState, RootState> & ADRActions = {
...
         for (let i = 0; i < uploadFilesPayload.length; i++) {
-            const { resourceType, resourceFilename, resourceId } = uploadFilesPayload[i]
+            const { resourceType, resourceFilename, resourceId, displayName } = uploadFilesPayload[i]
 
             const requestParams: Dict<string> = {resourceFileName: resourceFilename}
             if (resourceId){
                 requestParams["resourceId"] = resourceId
             }
 
+            if (displayName === "outputSummary") {
```
Comment (mwoodbri) [inline src/app/static/src/app/store/adr/actions.ts:162]: I know the strings happen to match here, but strictly speaking it might make more sense to compare against `resourceType` rather than `displayName` - the latter is really just an i18n tag.

File: src/app/static/src/app/types.ts [hunk 4]
```diff
@@ -149,4 +149,14 @@ export interface UploadFile {
     resourceId: string | null,
     resourceUrl: string | null,
     lastModified: string | null
+    description: string | null
```
Comment (mwoodbri) [inline src/app/static/src/app/types.ts:152]: Is this used anywhere?

File: src/app/static/src/app/store/modelCalibrate/actions.ts [hunk 5]
```diff
@@ -84,8 +84,8 @@ export const actions: ActionTree<ModelCalibrateState, RootState> & ModelCalibrat
                         },
                         {root: true});
                 }
-
                 commit(ModelCalibrateMutation.Calibrated);
+                commit({type: ModelCalibrateMutation.UploadMetadata, payload: data.uploadMetadata});
```
Comment (mwoodbri) [inline src/app/static/src/app/store/modelCalibrate/actions.ts:88]: Rather than adding an extra mutation that is only ever used here, I wonder if this should either just be part of `ModelCalibrateMutation.Calibrated` (if the uploadMetadata is genuinely integral to the calibration result) or an ADR mutation (if it's only related to uploads).

**Review by mwoodbri — CHANGES_REQUESTED** [review]
Looking good - just a couple of minor amendments and questions/clarifications

File: src/app/static/src/app/store/modelCalibrate/actions.ts (hunk 5 above)
Comment (EmmaLRussell) [inline src/app/static/src/app/store/modelCalibrate/actions.ts:88]: I don't think we need to commit the uploadMetadata separately at all - it's part of the main result set (`data`) which is being committed as part of modelRun/RunResultFetched - we could read it straight out of there when needed. We'd just need to add it to the CalibrateResultResponse type.

File: src/app/static/src/app/store/adr/actions.ts [hunk 6]
```diff
@@ -143,20 +144,28 @@ export const actions: ActionTree<ADRState, RootState> & ADRActions = {
...
+                requestParams["description"] = uploadMetadata.outputSummary.description
```
Comment (EmmaLRussell) [inline src/app/static/src/app/store/adr/actions.ts:163]: One issue with this approach that occurs to me is that it's not backward compatible - this will fail for any saved projects whose calibrate results were generated before this feature went in. I can imagine users wanting to upload files to ADR from existing projects. I guess we can tell them that they'll need to recalibrate in order to that? But in that case we should at least fail nicely / show them a meaningful error message which explains the problem. Any thoughts on that @r-ash ?

**Review by EmmaLRussell — COMMENTED** [review]
This isn't currently working for me, with latest hintr - getting error `Required String parameter 'description' is not present` - is this expected?

**Comment by carbowizzy** [comment]
> This isn't currently working for me, with latest hintr - getting error `Required String parameter 'description' is not present` - is this expected?
> 
> It seems that uploadMetadata has not been set in the modelCalibrate state, and it doesn't seem to be in the modelRun result - has it not been implemented in hintr yet? Are we waiting for this branch to be merged? [mrc-ide/hintr#266](https://github.com/mrc-ide/hintr/pull/266)
> If so, we should hold off merging this branch until then.

That is exactly my thought. It's not merged yet.

**Comment by mwoodbri** [comment]
> > This isn't currently working for me, with latest hintr - getting error `Required String parameter 'description' is not present` - is this expected?
> > It seems that uploadMetadata has not been set in the modelCalibrate state, and it doesn't seem to be in the modelRun result - has it not been implemented in hintr yet? Are we waiting for this branch to be merged? [mrc-ide/hintr#266](https://github.com/mrc-ide/hintr/pull/266)
> > If so, we should hold off merging this branch until then.
> 
> That is exactly my thought. It's not merged yet.

@carbowizzy To clarify that and to make it easier for reviewers to test it's best to put `mrc-2265` in [`hintr_version`](https://github.com/mrc-ide/hint/blob/master/src/config/hintr_version)

**Comment by carbowizzy** [comment]
> > > This isn't currently working for me, with latest hintr - getting error `Required String parameter 'description' is not present` - is this expected?
> > > It seems that uploadMetadata has not been set in the modelCalibrate state, and it doesn't seem to be in the modelRun result - has it not been implemented in hintr yet? Are we waiting for this branch to be merged? [mrc-ide/hintr#266](https://github.com/mrc-ide/hintr/pull/266)
> > > If so, we should hold off merging this branch until then.
> > 
> > 
> > That is exactly my thought. It's not merged yet.
> 
> @carbowizzy To clarify that and to make it easier for reviewers to test it's best to put `mrc-2265` in [`hintr_version`](https://github.com/mrc-ide/hint/blob/master/src/config/hintr_version)

@mwoodbri Yes hintr branch hasn't been merged. tbh I thought it was. The implementation works fine but we should wait for the implementation in hintr before merging.

File: src/app/static/src/app/store/adr/actions.ts (hunk 6 above)
Comment (mwoodbri) [reply src/app/static/src/app/store/adr/actions.ts:162]: I missed this, and my hunch is that we should prompt users to recalibrate rather than unilaterally upload some placeholder text that is ultimately inconsistent with the format/content of whatever we get from Naomi (which I know is still TBC). But agree that @r-ash should pass judgement here!

Comment (EmmaLRussell) [reply src/app/static/src/app/store/adr/actions.ts:162]: Agreed, should definitely not be uploading a placeholder, just need to check that prompting for recalibrate is acceptable - I think it should be as calibration is relatively quick, so that probably makes more sense than putting in some special workaround like extra endpoint just to fetch the description.

File: src/app/static/src/app/store/adr/actions.ts [hunk 7]
```diff
@@ -146,24 +145,30 @@ export const actions: ActionTree<ADRState, RootState> & ADRActions = {
...
-            if (displayName === "outputZip") {
+            } else if (uploadMetadata && resourceType === "outputZip") {
                 requestParams["description"] = uploadMetadata.outputZip.description
+            } else {
+                /**
+                 * Default description will tentatively solve backward
+                 * compatibility issue that may arise with previously calibrated model
+                 * and once re-calibrated, the correct data will be populated
+                 */
+                requestParams["description"] = "Naomi model outputs"
```
Comment (EmmaLRussell) [inline src/app/static/src/app/store/adr/actions.ts:171]: As @mwoodbri says, a better approach here would probably be to fail and to prompt the user to re-calibrate in error message. 

Also, a heads up that both this and some of the code in ADRController around description is going to conflict with the branch I'm currently working on to add upload of input files, where description is not a required parameter since for input resources description is not changed when new file version is uploaded (though again here I'm going to ask @r-ash if that's expected behaviour!)

**Review by EmmaLRussell — COMMENTED** [review]
Confirmed that this works as expected with hintr branch mrc-2265, and appends 'Naomi model outputs' as description - but sounds like we should talk to Rob on Monday about merging/deployment strategy.

**Review by EmmaLRussell — COMMENTED** [review]
I think you should ask @mwoodbri to re-review as well since he had comments last time.

I don't this is actually picking up the description from uploadMetadata in the caibrate result even though it should be as the hintr branch has merged. I've re-run deps since the merge, but I can't see `uploadMetadata` being returned in the calibrate result, hence the default description is what is always appended. Have you actually seen uploadMetadata in a calibrate result? Might need to take that up with Rob if it's not just something I've missed. 

Although the failing tests in buildkite are ADR tests, I think they should be investigated as do not seem to be typical 'ADR dev server being flaky' issues. The failing backend test is an unexpected 400, which has been consistent across a few builds. The failing frontend test is because model run result is not set, so that's certainly not an ADR problem.

File: src/app/static/src/app/store/adr/actions.ts [hunk 8]
```diff
@@ -154,9 +155,21 @@ export const actions: ActionTree<ADRState, RootState> & ADRActions = {
             const { resourceType, resourceFilename, resourceId } = uploadFilesPayload[i]
 
             const requestParams: Dict<string> = {resourceFileName: resourceFilename}
-            if (resourceId){
+            if (resourceId) {
                 requestParams["resourceId"] = resourceId
             }
+            if (uploadMetadata && resourceType === "inputs-unaids-naomi-report") {
```
Comment (EmmaLRussell) [inline src/app/static/src/app/store/adr/actions.ts:161]: These resource types shouldn't be hard-coded. Use state.schemas.outputZip and .outputSummary

**Review by EmmaLRussell — COMMENTED** [review]
From previous review
>I think you should ask @mwoodbri to re-review as well since he had comments last time.

>Although the failing tests in buildkite are ADR tests, I think they should be investigated as do not seem to be typical 'ADR dev server being flaky' issues. The failing backend test is an unexpected 400, which has been consistent across a few builds. The failing frontend test is because model run result is not set, so that's certainly not an ADR problem.

**Comment by EmmaLRussell** [comment]
Confirmed this works to upload meaningful description from hintr to ADR :+1:

File: NEWS.md [hunk 9]
```diff
@@ -10,6 +15,10 @@
 
 * Progress feedback on uploading files to ADR
 
+# hint 1.29.0
+
+* Include metadata when pushing file to ADR
+
```
Comment (mwoodbri) [inline NEWS.md:21]: Duplication

File: src/app/static/src/tests/adr/actions.test.ts [hunk 10]
```diff
@@ -1,17 +1,18 @@
...
-    mockModelCalibrateState,
+    mockModelCalibrateState, mockModelRunState,
     mockProjectsState,
     mockRootState,
     mockSuccess
 } from "../mocks";
 import {actions} from "../../app/store/adr/actions";
 import {ADRMutation} from "../../app/store/adr/mutations";
 import {UploadFile} from "../../app/types";
+import {CalibrateResultResponse} from "../../app/generated";
```
Comment (mwoodbri) [inline src/app/static/src/tests/adr/actions.test.ts:15]: This looks like an unnecessary change - please remove if so

**Review by mwoodbri — CHANGES_REQUESTED** [review]
Two tweaks then will be ready to merge

**Review by EmmaLRussell — CHANGES_REQUESTED** [review]
Sorry to punt this back again, but I really think you need to look into these test failures, which are still happening. As I said before:
>Although the failing tests in buildkite are ADR tests, I think they should be investigated as do not seem to be typical 'ADR dev server being flaky' issues. The failing backend test is an unexpected 400, which has been consistent across a few builds. The failing frontend test is because model run result is not set, so that's certainly not an ADR problem.

I don't want to merge to master with consistently failing tests..

File: src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/ADRController.kt [hunk 11]
```diff
@@ -167,42 +167,41 @@ class ADRController(private val encryption: Encryption,
...
     @PostMapping("/datasets/{id}/resource/{resourceType}/{modelCalibrateId}")
-    @Suppress("ReturnCount")
+    @Suppress("ReturnCount", "LongParameterList")
     fun pushFileToADR(@PathVariable id: String,
                       @PathVariable resourceType: String,
                       @PathVariable modelCalibrateId: String,
                       @RequestParam resourceFileName: String,
-                      @RequestParam resourceId: String?): ResponseEntity<String>
+                      @RequestParam resourceId: String?,
+                      @RequestParam description: String) : ResponseEntity<String>
```
Comment (mwoodbri) [inline src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/ADRController.kt:176]: `description` is now a required parameter, which may be the cause of the `ADRTests` failure.
However, rather than passing the description it might make more sense to retrieve `uploadMetadata` using `HintrAPIClient`. We do this for `downloadSpectrum` and `downloadSummary`, and `getPlottingMetadata` already exists, so maybe `getUploadMetadata` should also be added?

Comment (carbowizzy) [reply src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/ADRController.kt:176]: We spoke about getting the data from HintrAPIClient but resolved to having it in frontend instead because we will allow users to be able to edit/alter description on the dialog at some point.

Comment (mwoodbri) [reply src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/ADRController.kt:176]: Ah very good point - thanks

**Comment by carbowizzy** [comment]
> Sorry to punt this back again, but I really think you need to look into these test failures, which are still happening. As I said before:
> 
> > Although the failing tests in buildkite are ADR tests, I think they should be investigated as do not seem to be typical 'ADR dev server being flaky' issues. The failing backend test is an unexpected 400, which has been consistent across a few builds. The failing frontend test is because model run result is not set, so that's certainly not an ADR problem.
> 
> I don't want to merge to master with consistently failing tests..

**Comment by EmmaLRussell** [comment]
> > Sorry to punt this back again, but I really think you need to look into these test failures, which are still happening. As I said before:
> > > Although the failing tests in buildkite are ADR tests, I think they should be investigated as do not seem to be typical 'ADR dev server being flaky' issues. The failing backend test is an unexpected 400, which has been consistent across a few builds. The failing frontend test is because model run result is not set, so that's certainly not an ADR problem.
> > 
> > 
> > I don't want to merge to master with consistently failing tests..

? Did you intend to reply to this comment @carbowizzy ?

**Comment by carbowizzy** [comment]
> > > Sorry to punt this back again, but I really think you need to look into these test failures, which are still happening. As I said before:
> > > > Although the failing tests in buildkite are ADR tests, I think they should be investigated as do not seem to be typical 'ADR dev server being flaky' issues. The failing backend test is an unexpected 400, which has been consistent across a few builds. The failing frontend test is because model run result is not set, so that's certainly not an ADR problem.
> > > 
> > > 
> > > I don't want to merge to master with consistently failing tests..
> 
> ? Did you intend to reply to this comment @carbowizzy ?

Actually yes, the test failure was inherited from master. Since the failure is integration test then I believe there must be something wrong from ADR or config. This can be investigated in a different ticket.

**Comment by EmmaLRussell** [comment]
> Actually yes, the test failure was inherited from master. Since the failure is integration test then I believe there must be something wrong from ADR or config. This can be investigated in a different ticket.

These test failures are not inherited from master. Master build is passing. https://buildkite.com/mrc-ide/hint/builds?branch=master

Yes, these issues are in integration tests, but not all integration test failures are down to ADR issues. As I explained, if you look at the log for the front end test integration test failure you'll see that it is because modelRun is not set in the state, and it needs to be because of changes made in this branch. https://buildkite.com/mrc-ide/hint/builds/911#77e5bcc4-d81e-4462-89e2-b6f8f559e99f/206-553

I'm not sure what's causing the backend integration test failure. But given that master has been passing consistently and this branch has been failing, this merits investigation too before merging.

File: src/app/static/src/app/store/adr/actions.ts [hunk 12]
```diff
@@ -154,9 +155,21 @@ export const actions: ActionTree<ADRState, RootState> & ADRActions = {
...
+                requestParams["description"] = uploadMetadata.outputSummary.description
+            } else if (uploadMetadata && resourceType === state.schemas?.outputZip) {
+                requestParams["description"] = uploadMetadata.outputZip.description
+            } else {
+                /**
+                 * Default description will tentatively solve backward
+                 * compatibility issue that may arise with previously calibrated model
+                 * and once re-calibrated, the correct data will be populated
+                 */
+                requestParams["description"] = "Output upload from Naomi web"
```
Comment (mwoodbri) [inline src/app/static/src/app/store/adr/actions.ts:171]: Should this match [the Naomi boilerplate](https://github.com/mrc-ide/naomi/blob/master/R/run-model.R#L657-L664)? Note that it distinguishes between "output" and "summary report": we should probably do that, at least. For example:
```kotlin
if (resourceType === state.schemas?.outputSummary) {
   requestParams["description"] = uploadMetadata
        ? uploadMetadata.outputSummary.description
        : "Naomi summary report uploaded from Naomi web app"
}
```

**Review by mwoodbri — CHANGES_REQUESTED** [review]

**Review by mwoodbri — CHANGES_REQUESTED** [review]
One question about the default text that we're using.

You might be able to fix coverage by adding a test like this:
```kotlin
    @Test
    fun `pushes file to ADR fails if retrieval from hintr fails with empty response`()
    {
        val mockAPIClient: HintrAPIClient = mock {
            on { downloadSpectrum("model1") } doReturn ResponseEntity.status(HttpStatus.BAD_GATEWAY).body(null)
        }
        val sut = ADRController(mock(), mock(), mock(), mock(), mockProperties, mock(), mockAPIClient, mock(), mock())
        val result = sut.pushFileToADR("dataset1", "adr-output-zip", "model1", "output1.zip", "resource1", "Naomi model outputs")
        assertThat(result.statusCode).isEqualTo(HttpStatus.BAD_GATEWAY)
    }
```

File: src/app/static/src/app/store/adr/actions.ts (hunk 12 above)
Comment (carbowizzy) [reply src/app/static/src/app/store/adr/actions.ts]: I like the idea so I tried it at first but not working well. I will create a ticket to adopt it.

Comment (mwoodbri) [reply src/app/static/src/app/store/adr/actions.ts]: Do you remember what the problem was? Or have a link to ticket?

Even if this is just a fallback it would be better not to attach an identical description to both output files and summary reports. Ideally "Naomi output uploaded from Naomi web app" and "Naomi summary report uploaded from Naomi web app" respectively, for consistency with descriptions attached by Naomi.

Comment (carbowizzy) [reply src/app/static/src/app/store/adr/actions.ts]: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-2367

Comment (carbowizzy) [reply src/app/static/src/app/store/adr/actions.ts]: This is now updated in this ticket.

File: src/app/static/src/app/generated.d.ts [hunk 13]
```diff
@@ -109,6 +112,10 @@ export interface CalibrateResultResponse {
     upper: number | null;
     [k: string]: any;
   }[];
+  uploadMetadata: {
+    outputZip: UploadMetadataElements
+    outputSummary: UploadMetadataElements
+  } | null
```
Comment (mwoodbri) [inline src/app/static/src/app/generated.d.ts:118]: Was this change done manually or via `generate-types.sh`? See message at top of the file... manual changes will be overwritten.

**Review by mwoodbri — APPROVED** [review]
LGTM!

**Review by EmmaLRussell — APPROVED** [review]
Looks good, just needs conflicts fixing
